### DISCUSSION TITLE
dataset not register error when using visualizer for custom dataset

### DISCUSSION THREAD

**Discussion by suryasid09** [facebookresearch/detectron2#4795] [Q&A]
## Instructions To Reproduce the Issue:

I am using [detectron2](https://github.com/facebookresearch/detectron2)/[tools](https://github.com/facebookresearch/detectron2/tree/main/tools)/visualize_data.py with my custom dataset.
I am able to train and evaluate using my custom data. But, when I use visualize_data.py file, it throws an error that my dataset is not register. It also gives list of registered dataset and I don't see my custom dataset. 

1. Full runnable code or full changes you made:
```
Visulaization code: 
[detectron2](https://github.com/facebookresearch/detectron2)/[tools](https://github.com/facebookresearch/detectron2/tree/main/tools)/visualize_data.py 
There is no change made in this file, I am using the provided code as it is.

```
2. What exact command you run:

command format as in detectron2 github page: 
python visualize_data.py --config-file config.yaml --source annotation/dataloader --output-dir dir/ [--show]

The command used: 
!python3 /home/jovyan/thesis_s2577712/DeFRCN_voc_format/thesis-pascal_voc_format/defrcn/data/visualize_data.py --source annotation --config-file /home/jovyan/thesis_s2577712/DeFRCN_voc_format/thesis-pascal_voc_format/checkpoints/voc/runDeFrcn/defrcn_fsod_r101_novel1/fsrw-like/1shot_seed0_repeat0/config1.yaml --output-dir /home/jovyan/thesis_s2577712/DeFRCN_voc_format --show

4. __Full logs__ or other relevant observations:
```
Arguments: Namespace(config_file='/home/jovyan/thesis_s2577712/DeFRCN_voc_format/thesis-pascal_voc_format/checkpoints/voc/runDeFrcn/defrcn_fsod_r101_novel1/fsrw-like/1shot_seed0_repeat0/config1.yaml', opts=[], output_dir='/home/jovyan/thesis_s2577712/DeFRCN_voc_format', show=True, source='annotation')
Traceback (most recent call last):
  File "/home/jovyan/.local/lib/python3.8/site-packages/detectron2/data/catalog.py", line 51, in get
    f = self[name]
  File "/usr/lib/python3.8/collections/__init__.py", line 1010, in __getitem__
    raise KeyError(key)
KeyError: 'voc_2007_trainval_novel1_1shot_seed0'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/jovyan/thesis_s2577712/DeFRCN_voc_format/thesis-pascal_voc_format/defrcn/data/visualize_data.py", line 89, in <module>
    dicts = list(chain.from_iterable([DatasetCatalog.get(k) for k in cfg.DATASETS.TRAIN]))
  File "/home/jovyan/thesis_s2577712/DeFRCN_voc_format/thesis-pascal_voc_format/defrcn/data/visualize_data.py", line 89, in <listcomp>
    dicts = list(chain.from_iterable([DatasetCatalog.get(k) for k in cfg.DATASETS.TRAIN]))
  File "/home/jovyan/.local/lib/python3.8/site-packages/detectron2/data/catalog.py", line 53, in get
    raise KeyError(
KeyError: "Dataset 'voc_2007_trainval_novel1_1shot_seed0' is not registered! Available datasets are: coco_2014_train, coco_2014_val, coco_2014_minival, coco_2014_minival_100, coco_2014_valminusminival, coco_2017_train, coco_2017_val, coco_2017_test, coco_2017_test-dev, coco_2017_val_100, keypoints_coco_2014_train, keypoints_coco_2014_val, keypoints_coco_2014_minival, keypoints_coco_2014_valminusminival, keypoints_coco_2014_minival_100, keypoints_coco_2017_train, keypoints_coco_2017_val, keypoints_coco_2017_val_100, coco_2017_train_panoptic_separated, coco_2017_train_panoptic_stuffonly, coco_2017_train_panoptic, coco_2017_val_panoptic_separated, coco_2017_val_panoptic_stuffonly, coco_2017_val_panoptic, coco_2017_val_100_panoptic_separated, coco_2017_val_100_panoptic_stuffonly, coco_2017_val_100_panoptic, lvis_v1_train, lvis_v1_val, lvis_v1_test_dev, lvis_v1_test_challenge, lvis_v0.5_train, lvis_v0.5_val, lvis_v0.5_val_rand_100, lvis_v0.5_test, lvis_v0.5_train_cocofied, lvis_v0.5_val_cocofied, cityscapes_fine_instance_seg_train, cityscapes_fine_sem_seg_train, cityscapes_fine_instance_seg_val, cityscapes_fine_sem_seg_val, cityscapes_fine_instance_seg_test, cityscapes_fine_sem_seg_test, cityscapes_fine_panoptic_train, cityscapes_fine_panoptic_val, voc_2007_trainval, voc_2007_train, voc_2007_val, voc_2007_test, voc_2012_trainval, voc_2012_train, voc_2012_val, ade20k_sem_seg_train, ade20k_sem_seg_val"
​
```
## Expected behavior:
The model should have data registered, and it should visualize my custom dataset with bounding box on it.

## Environment:
PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2019.0.5 Product Build 20190808 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v1.5.0 (Git Hash e2ac1fac44c5078ca927cb9b90e1b3066a0b2ed0)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75
  - CuDNN 7.6.3
  - Magma 2.5.2
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DUSE_VULKAN_WRAPPER -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF,

**Comment by mark1ng123** [comment]
Hey there, when using custom datasets i would encourage you to write a visualiser of your own that way you can ensure that your datasets are in fact registered depending on which kind of dataset you use for example for coco its pretty straight forward:
```
from detectron2.data.datasets import register_coco_instances

register_coco_instances(training_dataset_name, {}, training_json, os.path.join(train_dataset_path, 'images'))
register_coco_instances(testing_dataset_name, {}, testing_json, os.path.join(test_dataset_path, 'images'))
```
when the training_dataset_name its the name you declared and the json is for the annotations, you can check it in the documentation.
If you used d2json for labelme2detectron2 its a bit different but also pretty simple using this piece of code: 
```
def convert_to_detectron_anottation(unique_labels):
    for d in ["train", "test"]:
        DatasetCatalog.register("ycb_" + d, lambda d=d: get_dataset_dicts(f"Dataset/{d}"))
        MetadataCatalog.get("ycb_" + d).set(thing_classes=unique_labels)


def get_dataset_dicts(img_dir):
    dataset_dicts = []
    img_dir = Path(img_dir)
    for i, jsonpath in enumerate([*img_dir.glob("*.d2json")]):
      with open(str(jsonpath), 'r') as fp:
        annot = json.load(fp)
        annot['image_id'] = i
        annot['file_name'] = str(img_dir/Path(annot['file_name']).name) 
        
        dataset_dicts.append(annot)
    
    return dataset_dicts
```
And for visualizing there are many examples online, but here is one:
```
def visualize_samples(ycb_metadata):
    train_dataset_dicts = get_dataset_dicts("Dataset/train")
    print(train_dataset_dicts)
    for d in random.sample(train_dataset_dicts, 3):    
        print(d["file_name"])
        img = cv2.imread(d["file_name"])
        visualizer = Visualizer(img[:, :, ::-1], metadata=ycb_metadata, scale=0.5)
        out = visualizer.draw_dataset_dict(d)
        plt.figure(figsize = (15,15))
        plt.imshow(out.get_image()[:, :, ::-1])
        plt.show()
```
and one for prediction:
```
def visualize_prediction(predictor, test_ds, metadata):
    test_dataset_dicts = test_ds
    for d in random.sample(test_dataset_dicts, 1):
        img = cv2.imread(d["file_name"])    
        outputs = predictor(img)
        instances = outputs["instances"].to("cpu")
        # Add bounding box and label name for debug
        v = Visualizer(img[:, :, ::-1], metadata=metadata,
                        scale=0.5, instance_mode=ColorMode.IMAGE_BW)
        for i, box in enumerate(instances.pred_boxes):
            v.draw_box(box)
        out = v.draw_instance_predictions(instances)
        plt.subplot(1,2,1)
        plt.imshow(out.get_image()[:, :, ::-1])
        plt.subplot(1,2,2)
        plt.imshow(img[:, :, ::-1])
        plt.show()
```
Try adjusting your custom visualizing to the format of your dataset :)

**Comment by suryasid09** [comment]
Thank you so much mark. This is a great help :)

**Comment by suryasid09** [comment]
I have a small question though! if I have registered dataset in another files, How can I use it in new visualization file without having to register it once again? I did try import but it doesn't seem to work.  I am new so apologies for this question.

**Comment by mark1ng123** [comment]
@suryasid09 you can try using jupyter notebook which saves all the kernel variables, but it dosent matter if you register it once again, the dataset is the same dataset as long as you provide the right paths in the functions.

**Comment by suryasid09** [comment]
> @suryasid09 you can try using jupyter notebook which saves all the kernel variables, but it dosent matter if you register it once again, the dataset is the same dataset as long as you provide the right paths in the functions.

@mark1ng123 thanks I will :)

**Comment by ppwwyyxx** [comment]
> f I have registered dataset in another files, How can I use it in new visualization file without having to register it once again

You can also import the "another file" from the "visualization file".

**Reply by suryasid09** [reply]
@ppwwyyxx Could you guide on how to do it may with some examples?  I did try importing, but every time it says that dataset is not registered. My dataset is pascal VOC format.
